File: backend/src/main/java/com/etekhno/rentacar/domain/BookingDTO.java
package com.etekhno.rentacar.domain;

import com.etekhno.rentacar.datamodel.enums.BookingStatus;
import com.etekhno.rentacar.domain.inbound.BookingDTOIn;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.math.BigDecimal;
import java.util.Date;

@Setter
@Getter
@NoArgsConstructor
@AllArgsConstructor
public class BookingDTO extends BookingDTOIn {
    private String id;
    public BookingDTO(String id, String vehicleId, String customerId, Date rentalStartDate, Date rentalEndDate,
                      BigDecimal totalAmount, BigDecimal amountPaid, BigDecimal remainingBalance, BookingStatus status) {
        super(vehicleId, customerId, rentalStartDate, rentalEndDate, totalAmount, amountPaid, remainingBalance, status);
        this.id = id;
    }
}

File: frontend/lib/domains/booking_dto.dart
import '/domains/enums/booking_status.dart';
import '/domains/inbound/booking_dto_in.dart';

class BookingDTO extends BookingDTOIn {
  String? id;

  BookingDTO({
    required this.id,
    required super.vehicleId,
    required super.customerId,
    required super.rentalStartDate,
    required super.rentalEndDate,
    required super.totalAmount,
    super.amountPaid = 0.0,
    super.remainingBalance,
    super.status = BookingStatus.PENDING,
  });

  factory BookingDTO.fromJson(Map<String, dynamic> json) {
    return BookingDTO(
      id: json['id'] as String?,
      vehicleId: json['vehicleId'] as String?,
      customerId: json['customerId'] as String?,
      rentalStartDate:
          DateTime.fromMillisecondsSinceEpoch(json['rentalStartDate'] as int),
      rentalEndDate:
          DateTime.fromMillisecondsSinceEpoch(json['rentalEndDate'] as int),
      totalAmount: (json['totalAmount'] as num?)?.toDouble(),
      amountPaid: (json['amountPaid'] as num?)?.toDouble() ?? 0.0,
      remainingBalance: (json['remainingBalance'] as num?)?.toDouble(),
      status: BookingDTOIn.parseBookingStatus(json['status']),
    );
  }

  @override
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'vehicleId': vehicleId,
      'customerId': customerId,
      'rentalStartDate': rentalStartDate?.toIso8601String(),
      'rentalEndDate': rentalEndDate?.toIso8601String(),
      'totalAmount': totalAmount,
      'amountPaid': amountPaid,
      'remainingBalance': remainingBalance,
      'status': status.toString().split('.').last,
    };
  }
}
